File: src/db/queries/symbols.ts
import { eq, sql } from "drizzle-orm";
import { db, schema } from "@/db";

const { symbols } = schema;

export interface SymbolMeta {
  symbol: string;
  name: string | null;
  currency: string;
  exchange: string | null;
}

export function getSymbolNames(): SymbolMeta[] {
  return db
    .select({
      symbol: symbols.symbol,
      name: symbols.name,
      currency: symbols.currency,
      exchange: symbols.exchange,
    })
    .from(symbols)
    .all();
}

export function getSymbolName(symbol: string): string | null {
  const row = db
    .select({ name: symbols.name })
    .from(symbols)
    .where(eq(symbols.symbol, symbol))
    .get();
  return row?.name ?? null;
}

export function updateSymbolName(
  symbol: string,
  name: string,
  exchange: string | null,
): void {
  db
    .update(symbols)
    .set({ name, exchange, updatedAt: sql`CURRENT_TIMESTAMP` })
    .where(eq(symbols.symbol, symbol))
    .run();
}

export function upsertSymbol(
  symbol: string,
  name: string | null,
  currency: string,
): void {
  db
    .insert(symbols)
    .values({ symbol, name, currency, updatedAt: sql`CURRENT_TIMESTAMP` })
    .onConflictDoUpdate({
      target: symbols.symbol,
      set: {
        // Keep the existing name if the new one is NULL.
        name: sql`COALESCE(excluded.name, ${symbols.name})`,
        currency: sql`excluded.currency`,
        updatedAt: sql`CURRENT_TIMESTAMP`,
      },
    })
    .run();
}

File: src/db/symbols.rs
use sqlx::{FromRow, SqlitePool};

#[derive(Debug, FromRow, Clone)]
pub struct Symbol {
    pub symbol: String,
    pub name: Option<String>,
    pub currency: String,
    pub exchange: Option<String>,
}

pub async fn get_symbol_names(pool: &SqlitePool) -> Result<Vec<Symbol>, sqlx::Error> {
    sqlx::query_as::<_, Symbol>("SELECT symbol, name, currency, exchange FROM symbols")
        .fetch_all(pool)
        .await
}

pub async fn get_symbol_name(
    pool: &SqlitePool,
    symbol: &str,
) -> Result<Option<String>, sqlx::Error> {
    let row: Option<(Option<String>,)> =
        sqlx::query_as("SELECT name FROM symbols WHERE symbol = ?")
            .bind(symbol)
            .fetch_optional(pool)
            .await?;
    Ok(row.and_then(|r| r.0))
}

pub async fn update_symbol_name(
    pool: &SqlitePool,
    symbol: &str,
    name: &str,
    exchange: Option<&str>,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        UPDATE symbols SET name = ?, exchange = ?, updated_at = CURRENT_TIMESTAMP
        WHERE symbol = ?
        "#,
    )
    .bind(name)
    .bind(exchange)
    .bind(symbol)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn upsert_symbol(
    pool: &SqlitePool,
    symbol: &str,
    name: Option<&str>,
    currency: &str,
) -> Result<(), sqlx::Error> {
    sqlx::query(
        r#"
        INSERT INTO symbols (symbol, name, currency, updated_at)
        VALUES (?, ?, ?, CURRENT_TIMESTAMP)
        ON CONFLICT(symbol) DO UPDATE SET
            name = COALESCE(excluded.name, symbols.name),
            currency = excluded.currency,
            updated_at = CURRENT_TIMESTAMP
        "#,
    )
    .bind(symbol)
    .bind(name)
    .bind(currency)
    .execute(pool)
    .await?;
    Ok(())
}
